src/NuGet.Clients/NuGetConsole.Host.PowerShell/PowerShellHost.cs

```diff
@@ -680,32 +685,80 @@ public string[] GetAvailableProjects()
 
             return NuGetUIThreadHelper.JoinableTaskFactory.Run(async delegate
                 {
-                    await NuGetUIThreadHelper.JoinableTaskFactory.SwitchToMainThreadAsync();
+                    var allProjects = await _solutionManager.GetNuGetProjectsAsync();
+
+                    var projectSafeNames = new List<string>();
+                    var displayNames = new List<string>();
+
+                    var safeNametasks = new List<Task<string>>();
+                    var displayNameTasks = new List<Task<string>>();
+
+                    foreach (var project in allProjects)
+                    {
+                        // Throttle and wait for a task to finish if we have hit the limit
+                        if (safeNametasks.Count == MaxTasks)
+                        {
+                            var safeName = await CompleteTaskAsync(safeNametasks);
+                            projectSafeNames.Add(safeName);
+                        }
+
+                        if (displayNameTasks.Count == MaxTasks)
+                        {
+                            var displayName = await CompleteTaskAsync(displayNameTasks);
+                            displayNames.Add(displayName);
+                        }
 
-                    var allProjects = _solutionManager.GetNuGetProjects();
-                    _projectSafeNames = allProjects.Select(_solutionManager.GetNuGetProjectSafeName).ToArray();
-                    var displayNames = allProjects.Select(GetDisplayName).ToArray();
-                    Array.Sort(displayNames, _projectSafeNames, StringComparer.CurrentCultureIgnoreCase);
+                        var safeNameTask = Task.Run(async () => await _solutionManager.GetNuGetProjectSafeNameAsync(project));
```

Comment:
This is to improve performance when you click on project list combo box on PMC. With a decent solution size, it takes several seconds to come up and hang VS. 

These tasks are not UI bound since getting NuGet project safe name doesn't need to run on UI. This is the simplest and yet efficient logic I could come up. Besides we are already using similar strategy in our existing code like for restore tasks. But please feel free to suggest any better, efficient logic.